hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java

```diff
@@ -2312,23 +2312,23 @@ private void invalidate(String bpid, Block[] invalidBlks, boolean async)
     for (int i = 0; i < invalidBlks.length; i++) {
       final ReplicaInfo removing;
       final FsVolumeImpl v;
-      try (AutoCloseableLock lock = lockManager.writeLock(LockLevel.BLOCK_POOl, bpid)) {
-        final ReplicaInfo info = volumeMap.get(bpid, invalidBlks[i]);
-        if (info == null) {
-          ReplicaInfo infoByBlockId =
-              volumeMap.get(bpid, invalidBlks[i].getBlockId());
-          if (infoByBlockId == null) {
-            // It is okay if the block is not found -- it
-            // may be deleted earlier.
-            LOG.info("Failed to delete replica " + invalidBlks[i]
-                + ": ReplicaInfo not found.");
-          } else {
-            errors.add("Failed to delete replica " + invalidBlks[i]
-                + ": GenerationStamp not matched, existing replica is "
-                + Block.toString(infoByBlockId));
-          }
-          continue;
+      final ReplicaInfo info = volumeMap.get(bpid, invalidBlks[i]);
+      if (info == null) {
+        ReplicaInfo infoByBlockId =
+                volumeMap.get(bpid, invalidBlks[i].getBlockId());
+        if (infoByBlockId == null) {
+          // It is okay if the block is not found -- it
+          // may be deleted earlier.
+          LOG.info("Failed to delete replica " + invalidBlks[i]
+                  + ": ReplicaInfo not found.");
+        } else {
+          errors.add("Failed to delete replica " + invalidBlks[i]
+                  + ": GenerationStamp not matched, existing replica is "
+                  + Block.toString(infoByBlockId));
         }
+        continue;
+      }
+      try (AutoCloseableLock lock = lockManager.writeLock(LockLevel.VOLUME, bpid, info.getStorageUuid())) {
```

Comment:
When there are many replicas should be deleted, there will be more lock competition, so I think we could optimize it here. What do you think? cc @MingXiangLi any suggestions?

Reply:
This is the reason I acquire block pool lock before...Or we sort these block and batch processing with the same volume?